src/RefImplementations.cc

```diff
@@ -1178,16 +1247,75 @@ int rowwise_sparse_adagrad_fused_ref(
       }
 
       float* h_ = h + idx;
-      float* w_ = w + idx * block_size;
+      DataType* w_ = w + idx * block_size;
 
       float hi = *h_ = *h_ + final_sum;
       float float_step = lr / (std::sqrt(hi) + epsilon);
 
-      for (int j = 0; j < block_size; ++j) {
-        w_[j] += g_[j] * float_step;
+      int nvec = (block_size + vlen - 1) / vlen;
+      int rem = (block_size % vlen) ? (block_size % vlen) : vlen;
+
+      // Emulate JIT behavior of stochastic rounding with vector-length
+      //
+      // Generate R buffer every 4 steps of nvec loop. Each 8-bit in R
+      // (uint32_t) will be used once. It is shifted to bits[5..13] then
+      // added to FP32 weights before FP16 conversion.
+      //
+      // The shifted 8 bit region
+      // +-------+--------+--------+--------+
+      // |       |        |   xxxxx|xxx     |
+      //  31      23       15       7      0
+      //
+      // Half float has 10 bits of mantissa, and float has 23, we are shifting
+      // the bits to cover the region where half floats can't represent data.
+      // This is bit 13-23 of the mantissa of fp32.
+      // This will be effectively adding a random variable of [0,1]
+
+      for (int n = 0; n < nvec; ++n) {
+        int len = (n == nvec - 1) ? rem : vlen;
+        int sr_idx = n % 4;
+
+        if (isFloat16w && use_stochastic_rounding) {
+          if (sr_idx == 0) {
+            for (int v = 0; v < vlen; ++v) {
+              R[v] = rnd128_next(v, vlen);
+              r[v] = (R[v] << 24) >> 19;
+            }
+          } else if (sr_idx == 1) {
+            for (int v = 0; v < vlen; ++v)
+              r[v] = ((R[v] >> 8) << 24) >> 19;
```

Comment:
Maybe we can make shifting logics more consistent?
for sr_idx == 1
(R[v] << 16) >> 19
for sr_idx == 2
(R[v] << 8) >> 19
for sr_idx == 3
R[v] >> 19

Reply:
Yes it can be so but it may need an additional mask to clear other not used bits. But you are right we can still make it more consistent in REF path. Perhaps we can just use the equivalent logic below:

```
case sr_idx:
0: (R[v] && 0xFFU) << 5
1: (R[v] & 0xFF00U) >> 3
2: (R[v] & 0xFF0000U) >> 11
3: (R[v] & 0xFF000000U) >> 19
```

In JIT path the above logic requires additional vector registers (and additional mask loads in AVX2) this is why the shift is used.